src/RankedPairsGraph.php

```diff
@@ -0,0 +1,108 @@
+<?php
+namespace PivotLibre\Tideman;
+
+use \Fhaculty\Graph\Graph;
+use \Fhaculty\Graph\Vertex;
+use \Fhaculty\Graph\Set\Vertices;
+// use \Fhaculty\Graph\Edge;
+use \Graphp\Algorithms\Search\DepthFirst;
+use Graphp\Algorithms\ShortestPath\BreadthFirst;
+
+class RankedPairsGraph
+{
+    private $graph;
+    //vertices have named attributes that we can associate artbitrary data to. This constant defines the name of an
+    //attribute that we will use to store the original Candidate object that a vertex corresponds to.
+    private const CANDIDATE_ATTRIBUTE_NAME = "candidate";
+
+    public function __construct()
+    {
+        $this->graph = new Graph();
+    }
+
+    /**
+    * @return CandidateList - a list of Candidates in descending order of preference. Candidates that are more
+    * preferred have a lower index than Candidates that are less preferred.
+    */
+    public function getWinningCandidates() : CandidateList
+    {
+        $winningVertices = $this->getSourceVertices($this->graph);
+        $tiedWinners = $this->getCandidatesFromVertices($winningVertices);
+        return $tiedWinners;
+    }
+    /**
+     * Locks in Margins in order of descending difference, ignoring any Margins that would contradict a
+     * previously-locked-in Margin.
+     * @param MarginList a MarginList whose Margins are sorted in order of descending difference and all differences are
+     * greater than or equal to zero.
+     */
+    public function addMargins(MarginList $sortedMarginList) : CandidateList
+    {
+        foreach ($sortedMarginList as $margin) {
+            $this->addMargin($margin);
+        }
+    }
+    public function addMargin(Margin $margin)
+    {
+        $winnerVertex = $this->addCandidateToGraph($margin->getWinner(), $this->graph);
+        $loserVertex = $this->addCandidateToGraph($margin->getLoser(), $this->graph);
+        $newEdge = $winnerVertex->createEdgeTo($loserVertex);
+        $newEdge->setWeight($margin->getDifference());
+        //check for contradiction of stronger preferences
+        if ($this->vertexIsInACycle($loserVertex)) {
+            //don't contradict stronger preferences that have been locked in earlier
+            $newEdge->destroy();
+        }
+    }
+
+    /**
+    * @todo #8 optimize this so that it uses a depth first search that returns early if the source node is discovered.
+    */
+    public function vertexIsInACycle(Vertex $vertex) : bool
```

Comment:
@tylerharter do you think that this approach would yield correct cycle detection?
I realize it could be more efficient. I'm using the library's canned functionality for now.
https://github.com/graphp/algorithms/blob/81db4049c35730767ec8f97fb5c4844234b86cef/src/ShortestPath/BreadthFirst.php